Spigot-API-Patches/0232-Add-plugin-reload-policy.patch

```diff
@@ -0,0 +1,188 @@
++     * <li>Plugins will not follow this policy if Bukkit is told to ignore them.
++     * </ul>
++     * <p>
++     * In the plugin.yml, this entry is named <code>reload</code>.
++     * <p>
++     * Example: <blockquote><pre>reload: always</pre></blockquote>
++     *
++     * @return the plugin's reload policy
++     */
++    @NotNull
++    public PluginReloadPolicy getReloadPolicy() {
++        return reloadPolicy;
++    }
++    // Paper end
++
+     /**
+      * @return unused
+      * @deprecated unused
+@@ -1127,6 +1150,16 @@ public final class PluginDescriptionFile {
+             apiVersion = map.get("api-version").toString();
+         }
+ 
++        // Paper start - Add reload policy
++        if (map.get("reload") != null) {
++            PluginReloadPolicy policy = PluginReloadPolicy.getByName(map.get("reload").toString());
++            if (policy == null) {
++                throw new InvalidDescriptionException("reload is not a valid choice");
++            }
++            reloadPolicy = policy;
++        }
++        // Paper end
++
+         try {
+             lazyPermissions = (Map<?, ?>) map.get("permissions");
+         } catch (ClassCastException ex) {
+@@ -1201,6 +1234,8 @@ public final class PluginDescriptionFile {
+             map.put("api-version", apiVersion);
+         }
+ 
++        map.put("reload", reloadPolicy); // Paper - Add reload policy
++
+         if (classLoaderOf != null) {
+             map.put("class-loader-of", classLoaderOf);
+         }
+diff --git a/src/main/java/org/bukkit/plugin/PluginReloadPolicy.java b/src/main/java/org/bukkit/plugin/PluginReloadPolicy.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..2d54564864182a73295ac28f9cf089b2c079230e
+--- /dev/null
++++ b/src/main/java/org/bukkit/plugin/PluginReloadPolicy.java
+@@ -0,0 +1,40 @@
++package org.bukkit.plugin;
++
++import org.jetbrains.annotations.NotNull;
++
++import java.util.Locale;
++
++/**
++ * Represents the behavior of a plugin when reloaded. Note that these policies
++ * only apply if the server isn't reloaded using the legacy method.
++ */
++public enum PluginReloadPolicy {
++
++    /**
++     * The plugin will always be reloaded assuming its dependencies are also reloadable.
++     */
++    ALWAYS,
++    /**
++     * The plugin will never be reloaded.
++     */
++    NEVER,
++    /**
++     * The plugin will always be reloaded assuming the plugin's file hash hasn't changed
++     * and its dependencies are also reloadable.
++     */
++    SAME_HASH;
++
++    /**
++     * Looks up a PluginReloadPolicy by possible plugin.yml values.
++     *
++     * @param name Name of the policy.
++     * @return Specified value, or null if not found.
++     */
++    public static PluginReloadPolicy getByName(@NotNull String name) {
++        try {
++            return valueOf(name.toUpperCase(Locale.ENGLISH).replaceAll("[^A-Z-]", "").replace("-", "_"));
```

Comment:
```suggestion
+            return valueOf(name.toUpperCase(Locale.ENGLISH).replace("-", "_"));
```

Don't correct the user; them entering e.g. `123` is not grounds for Paper to care as that's their fault. The replacing `-` -> `_` is however alright as that's a very easy thing to fuck up (*cough* Bungee & depend/-s...).